### PR TITLE
Make all externals build for 4.3 (main)

### PR DESCRIPTION
**Pull request by korydraughn**
Changes resulting in ignoring warnings will be fixed once all externals build without error.

### CODE REVIEW

File: versions.json [hunk 1]
```diff
@@ -292,14 +295,15 @@
         "fpm_directories": ["bin","include","lib"]
     },
     "mungefs": {
-        "commitish": "1.0.4",
+        "github_repo": "https://github.com/korydraughn/mungefs",
+        "commitish": "4.3_compat",
```
Comment (korydraughn) [inline versions.json:299]: merged mungefs fixes.

these lines can be removed.

File: versions.json [hunk 2]
```diff
@@ -342,16 +346,16 @@
         ],
         "fpm_directories": ["include", "lib"]
     },
-    "qpid": {
-        "commitish": "qpid-cpp-0.34",
-        "version_string": "0.34",
+    "qpid-cpp": {
```
Comment (korydraughn) [inline versions.json:349]: does not compile against python 3.

File: versions.json [hunk 3]
```diff
@@ -35,8 +35,9 @@
         "fpm_directories": ["bin","include","lib"]
     },
     "aws-sdk-cpp": {
-        "commitish": "1.4.89",
-        "version_string": "1.4.89",
+        "github_repo": "https://github.com/aws/aws-sdk-cpp",
+        "commitish": "1.9.185",
+        "version_string": "1.9.185",
```
Comment (korydraughn) [inline versions.json:40]: requires pulling in latest changes from upstream.

File: versions.json [hunk 4]
```diff
@@ -173,15 +175,16 @@
...
+        "commitish": "d68e30e382b5f2817be8cd901494736b26d4896e",
+        "version_string": "d68e30e3",
         "license": "MIT",
-        "consortium_build_number": "2",
+        "consortium_build_number": "0",
         "externals_root": "opt/irods-externals",
         "build_steps": [
-            "git submodule update --init --recursive",
+            "cd external/cpr; git checkout 1.7.2",
+            "cd external/jsoncpp; git checkout 1.9.5",
```
Comment (korydraughn) [inline versions.json:185]: checks out latest stable releases for submodules.

File: versions.json [hunk 5]
```diff
@@ -173,15 +175,16 @@
...
+            "cd build; rm -f CMakeCache.txt;TEMPLATE_CMAKE_EXECUTABLE -G 'Unix Makefiles' -DCMAKE_CXX_COMPILER=TEMPLATE_CLANGPP_EXECUTABLE -DCMAKE_C_COMPILER=TEMPLATE_CLANG_EXECUTABLE -DCMAKE_INSTALL_RPATH=/TEMPLATE_BOOST_RPATH\\;/TEMPLATE_CLANG_RUNTIME_RPATH\\;\\$ORIGIN\\;/TEMPLATE_CPR_RPATH -DCMAKE_INSTALL_PREFIX=TEMPLATE_INSTALL_PREFIX -DCMAKE_INSTALL_LIBDIR=lib -DBOOST_ROOT=TEMPLATE_BOOST_ROOT -DCMAKE_CXX_FLAGS='-std=c++14 -nostdinc++ -Wno-unused-command-line-argument -ITEMPLATE_CLANG_CPP_HEADERS' -DCMAKE_EXE_LINKER_FLAGS='-stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi -lpthread' -DCMAKE_SHARED_LINKER_FLAGS='-Wl,--enable-new-dtags -Wl,-z,origin -stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi' -DCMAKE_MODULE_LINKER_FLAGS='-stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi' ..",
```
Comment (korydraughn) [inline versions.json:187]: try to remove use `-Wno-*` compiler options

File: versions.json [hunk 6]
```diff
@@ -292,14 +295,15 @@
...
+            "cd build; rm -f CMakeCache.txt;TEMPLATE_CMAKE_EXECUTABLE -G 'Unix Makefiles' -DCPPZMQ_PATH=TEMPLATE_CPPZMQ_PATH -DZMQ_PATH=TEMPLATE_ZMQ_PATH -DAVRO_PATH=TEMPLATE_AVRO_PATH -DCLANG_LIBS=TEMPLATE_CLANG_CPP_LIBRARIES -DBOOST_PATH=TEMPLATE_BOOST_ROOT -DCMAKE_INSTALL_PREFIX=TEMPLATE_INSTALL_PREFIX -DCMAKE_CXX_COMPILER=TEMPLATE_CLANGPP_EXECUTABLE -DCMAKE_C_COMPILER=TEMPLATE_CLANG_EXECUTABLE -DCMAKE_CXX_FLAGS='-std=c++14 -nostdinc++ -Wno-deprecated-declarations -ITEMPLATE_CLANG_CPP_HEADERS' -DCMAKE_EXE_LINKER_FLAGS='-stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi' -DCMAKE_SHARED_LINKER_FLAGS='-stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi' -DCMAKE_MODULE_LINKER_FLAGS='-stdlib=libc++ -LTEMPLATE_CLANG_CPP_LIBRARIES -lc++abi' -DIRODS_EXTERNALS_PACKAGE_ROOT=TEMPLATE_SCRIPT_PATH -DCMAKE_INSTALL_RPATH=/TEMPLATE_BOOST_RPATH\\;/TEMPLATE_AVRO_RPATH\\;/TEMPLATE_ZMQ_RPATH\\;/TEMPLATE_LIBARCHIVE_RPATH\\;/TEMPLATE_CLANG_RUNTIME_RPATH ..",
```
Comment (korydraughn) [inline versions.json:306]: try to remove use `-Wno-*` compiler options

File: versions.json [hunk 7]
```diff
@@ -342,16 +346,16 @@
         ],
         "fpm_directories": ["include", "lib"]
     },
-    "qpid": {
-        "commitish": "qpid-cpp-0.34",
-        "version_string": "0.34",
+    "qpid-cpp": {
+        "github_repo": "https://github.com/apache/qpid-cpp",
+        "commitish": "1.39.0",
+        "version_string": "1.39.0",
```
Comment (korydraughn) [inline versions.json:352]: requires pulling in latest changes from upstream.

File: versions.json [hunk 8]
```diff
@@ -360,14 +364,15 @@
         "fpm_directories": []
     },
     "qpid-proton": {
-        "commitish": "0.23.0",
-        "version_string": "0.23.0",
+        "github_repo": "https://github.com/apache/qpid-proton",
+        "commitish": "0.36.0",
+        "version_string": "0.36.0",
```
Comment (korydraughn) [inline versions.json:369]: requires pulling in latest changes from upstream.

File: Makefile [hunk 9]
```diff
@@ -190,7 +190,7 @@ qpid-proton_clean :
 	@rm -rf qpid-proton*
 	@rm -rf $(QPID-PROTON_PACKAGE)
 
-$(QPID-WITH-PROTON_PACKAGE) : $(QPID_PACKAGE)
+$(QPID-WITH-PROTON_PACKAGE) : $(QPID-CPP_PACKAGE)
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply Makefile]: removed qpid-cpp and qpid-with-proton

File: versions.json (hunk 7 above)
Comment (korydraughn) [reply versions.json]: removed qpid-cpp and qpid-with-proton

File: versions.json (hunk 2 above)
Comment (korydraughn) [reply versions.json]: removed qpid-cpp and qpid-with-proton

File: versions.json (hunk 3 above)
Comment (trel) [reply versions.json:40]: i've updated irods/aws-sdk-cpp, it's now in sync with upstream main.

File: versions.json (hunk 8 above)
Comment (trel) [reply versions.json:352]: i've updated irods/qpid-proton, it's now in sync with upstream main.

File: build.py [hunk 10]
```diff
@@ -263,22 +253,14 @@ def build_package(target, build_native_package):
...
-        os.chdir(os.path.join(build_dir,target))
-        run_cmd(['git', 'fetch'], check_rc='git fetch failed')
-        run_cmd(['git', 'checkout', v['commitish']], check_rc='git checkout failed')
+    elif not os.path.isdir(os.path.join(build_dir,target)):
+        mkdir_p(build_dir)
+        os.chdir(build_dir)
+        log.debug('cwd: {0}'.format(os.getcwd()))
+        git_cmd = ['git', 'clone', '--depth', '1', '--branch', v['commitish'], '--recurse-submodules'] 
+        git_cmd.append(v['github_repo'] if 'github_repo' in v else 'https://github.com/irods/{0}'.format(target))
+        run_cmd(git_cmd, check_rc='git checkout failed')
```
Comment (trel) [inline build.py:262]: git clone failed

File: versions.json (hunk 1 above)
Comment (korydraughn) [reply versions.json]: done

File: versions.json (hunk 3 above)
Comment (korydraughn) [reply versions.json]: removed `"github_repo"`

File: versions.json (hunk 8 above)
Comment (korydraughn) [reply versions.json]: removed `"github_repo"`

**Comment by trel** [comment]
the naming of 'github_repo'... needs work.

It's not necessarily github... could be anywhere that git can be pointed to, including a local directory...

and repo... ew.

So, consider renaming 'github_repo' to 'git_repository' or 'alternative_git_repository'.

I'm leaning towards 'git_repository'...

File: build.py (hunk 10 above)
Comment (korydraughn) [reply build.py]: fixed

File: build.py [hunk 11]
```diff
@@ -263,22 +253,14 @@ def build_package(target, build_native_package):
...
-            log.debug('cwd: {0}'.format(os.getcwd()))
-            if target == 'boost':
-                # using boostorg namespace instead of forking 50+ relatively linked submodules
-                run_cmd(['git', 'clone', 'https://github.com/boostorg/boost'])
-            else:
-                run_cmd(['git', 'clone', 'https://github.com/irods/{0}'.format(target)])
-        os.chdir(os.path.join(build_dir,target))
-        run_cmd(['git', 'fetch'], check_rc='git fetch failed')
-        run_cmd(['git', 'checkout', v['commitish']], check_rc='git checkout failed')
+    elif not os.path.isdir(os.path.join(build_dir,target)):
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply build.py]: moved into else case.

the condition is still checked. this PR removes the ability to checkout a different commit following a build error. if the dev wants to use a different commit, they will need to run `make <target>_clean` first.

File: versions.json (hunk 4 above)
Comment (korydraughn) [reply versions.json:184]: this requires making the change in our forked version or submitting changes to the upstream repo.

this keeps us from having to do either. i'm not advocating for this, but i can say this solves the issue without having to rely on any hard code changes.

**Comment by korydraughn** [comment]
git_repository is good.

File: README.md [hunk 12]
```diff
@@ -60,3 +60,10 @@ dnf update -y && dnf install -y sudo git python3 python3-distro gcc-toolset-11
 scl enable gcc-toolset-11 bash
 make # or "make server" for packages specific to building the iRODS server.
 ```
+
+# FAQ
+
+### Q. Can I build an externals package using a different repository?
+Yes. For the externals package of interest, simply set the `"github_repo"` property to the Github URL of repository to clone from. If this property is not present, it means the build will clone the forked repository under https://github.com/irods. The build expects that the clone will have a directory name matching the name of the forked project under https://github.com/irods.
+
+Defining a different Github repository to use for an external normally means a change in the version. For that reason, it is important to remember that `"commitish"` is required to match a branch name or tag. SHA's do not work.
```
Comment (korydraughn) [inline README.md:69]: @trel @alanking thoughts?

File: README.md [hunk 13]
```diff
@@ -64,6 +64,6 @@ make # or "make server" for packages specific to building the iRODS server.
...
+Yes. For the externals package of interest, simply set the `"git_repository"` property to the URL of the repository to clone. If this property is not present, it means the build will clone the forked repository under https://github.com/irods. The build expects that the clone will have a directory name matching the name of the forked project under https://github.com/irods.
 
-Defining a different Github repository to use for an external normally means a change in the version. For that reason, it is important to remember that `"commitish"` is required to match a branch name or tag. SHA's do not work.
+Defining a different git repository to use for an external normally means a change in the version. For that reason, it is important to remember that `"commitish"` is required to match a **branch name** or **tag**. SHA's do not work.
```
Comment (trel) [inline README.md:69]: > SHA's do not work

is this true?

Comment (korydraughn) [reply README.md:69]: yes. i'm adding a knob that allows a SHA to be passed.

this is due to the use of cloning a specific branch via the --branch option.

Comment (korydraughn) [reply README.md]: we need a good name for the new option. the current PR introduces "using_sha" as the name of the option.

the name correctly captures the primary reason for its use, but it isn't limited to one usage pattern. what it really does is separate the fetch/checkout step from the clone step.

suggestions:
- `execute_fetch_checkout_separately`
- `fetch_checkout_separately`
- `using_sha`
- `commitish_is_sha` (this is horrible)
- `clone_separately` (this is horrible too)

Comment (trel) [reply README.md]: oooh, i didn't realize `--branch` didn't allow SHAs.   interesting.

Comment (trel) [reply README.md]: what if we always just do the clone, then cd, then checkout.     and skip using the `--branch`?

then no additional option needed.

Comment (korydraughn) [reply README.md]: sure we can do that. the reason it was added was so we cloned only the things related to that specific commit, which is really only a problem when handling large repos with a lot of git history.

Comment (korydraughn) [reply README.md]: do we want to go back to the old behavior?

**Comment by korydraughn** [comment]
as of commit 944b7ca, all externals build successfully.

all that is left is to resolve the remaining comments.

File: README.md (hunk 13 above)
Comment (trel) [reply README.md]: Ah, I see.    That does seem worth it...

How about...

`use_specific_sha`
or just
`use_sha`

Comment (korydraughn) [reply README.md]: i like it

Comment (korydraughn) [reply README.md]: renamed to `enable_sha`.

File: versions.json (hunk 5 above)
Comment (korydraughn) [reply versions.json]: removed.

File: versions.json (hunk 6 above)
Comment (korydraughn) [reply versions.json:307]: the compilation errors are due to use of deprecated code. deprecation does not mean the code cannot be used, but that it is to be removed in the future.

we can get rid of the compilation errors by using the new APIs. if we fix mungefs, we will need to decide whether the version number needs to be bumped, or we use the SHA.

given that mungefs is really a testing tool, fixing the mungefs code and/or disabling the warnings via compiler option is totally fine.

File: README.md [hunk 14]
```diff
@@ -60,3 +60,22 @@ dnf update -y && dnf install -y sudo git python3 python3-distro gcc-toolset-11
 scl enable gcc-toolset-11 bash
 make # or "make server" for packages specific to building the iRODS server.
 ```
+
+# FAQ
+
+### Q. Can I build an externals package using a different repository?
+Yes. For the externals package of interest, simply set the `"git_repository"` property to the URL of the repository to clone. If this property is not present, it means the build will clone the forked repository under https://github.com/irods. The build expects that the clone will have a directory name matching the name of the forked project under https://github.com/irods.
```
Comment (alanking) [inline README.md:67]: Should we mention `versions.json` to make it clear where the property being set is located? I didn't pick up on that the first time reading this

**Review by alanking — COMMENTED** [review]
Lookin good

File: README.md (hunk 14 above)
Comment (korydraughn) [reply README.md:67]: good point. will add that in.

Comment (korydraughn) [reply README.md]: updated.

thoughts?

Comment (alanking) [reply README.md]: Looks good to me!

File: README.md [hunk 15]
```diff
@@ -64,9 +64,9 @@ make # or "make server" for packages specific to building the iRODS server.
...
 
-Defining a different git repository to use for an external normally means a change in the version. For that reason, it is important to remember that `"commitish"` is required to match a **branch name** or **tag**. If you want to use a SHA for the `"commitish"`, you'll need to set `"enable_sha"` to true. This instructs the build system to separate the fetching/checking-out of a commit from the cloning of the repository. To better explain what this means, see below.
+Defining a different git repository to use for an external normally means a change in the version. For that reason, it is important to remember that `"commitish"` is required to match a **branch name** or **tag**. If you want to use a **SHA** for the `"commitish"`, you'll need to set `"enable_sha"` to true. This instructs the build system to separate the fetching/checking-out of a commit from the cloning of the repository. To better explain what this means, see below.
```
Comment (trel) [inline README.md:69]: use `true`, so it looks like a keyword/value?   here and on line 71 (`false`) and 75 (`true`).

Comment (korydraughn) [reply README.md]: done

**Comment by trel** [comment]
ship it.

**Comment by trel** [comment]
got issue numbers for all these first?  just in case?

**Comment by korydraughn** [comment]
i'll create those too.

**Comment by korydraughn** [comment]
all packages were built successfully.

added #'s

**Pull request merged**